### ISSUE TITLE
Callbacks

### ISSUE THREAD

**Issue by simonbyrne** [JuliaStats/RCall.jl#58]
It would be useful to be able to call Julia functions from within R, similar to PyCall's facility, e.g.
```julia
rcall(:plot, x -> x*x)
```
However there doesn't appear to be an immediately obvious way to call C function pointers from R, at least as far as I could tell. Any ideas?

**Comment by dmbates** [comment]
AFAIK the only way of calling C functions from R is to write and compile wrappers.  If all your callbacks had a fixed signature you could create an R function that called C/C++ code that takes an external pointer and sets up the call.  Rather messy but I think that is about the best that can be done.  The Rcpp package and the inline package for R have ways of doing the compilation on the fly.

**Comment by simonbyrne** [comment]
Having dug around some of the undocumented headers, I think I may have found a way to do this.

**Comment by simonbyrne** [comment]
@dmbates I'm trying to understand more about the internals: why do you manually reassign `vv.p` and `vv.pv` [here](https://github.com/JuliaStats/RCall.jl/blob/3b1f9bc8257b6fa45e077bbabd5a3f7d3dd6da25/src/sexp.jl#L24-L25)? Doesn't this break the doubly-linked list used for R's memory management?

**Comment by dmbates** [comment]
The `unsafe_load` call allocates a Julia vector and copies the values pointed to by `p` into the Julia memory space.  It is the elements of that vector that are being overwritten, not the original memory locations allocated by R.

**Comment by simonbyrne** [comment]
Ah, I has just reached that conclusion. Thanks!

**Comment by dmbates** [comment]
My comment was poorly phrased.  There are two calls to `unsafe_load` and I am referring to the second one.  
```jl
vv = unsafe_load(convert(Ptr{jtyp},p))
```

Also, the values are in a Julia type, not a vector.

**Comment by randy3k** [comment]
I have implemented it in RCalling.jl, which is obviously inspired by rpy2.

Basically, we want to convect a julia function to a R function. It could be done in the following steps.

1. register an external method via R api (`do_julia` in my code) which takes julia objects (the function to be called and the arguments) and eval the call. This will allow us to call `.External(".RCall")` in R. 
Check https://github.com/randy3k/RCalling.jl/blob/master/deps/external.c

2. covert the input julia function call to the R function-call object (via `rlang`): `function(...) {.External(".RCall", POINTER, ...)}`
where POINTER is the pointer to the array of julia objects (the julia function to be called and the arguments).

**Comment by randy3k** [comment]
I may have some time tomorrow to work on this.

**Comment by simonbyrne** [comment]
I actually have some code as part of my other PR I'm going to put up soon, after a bit more testing.

**Comment by randy3k** [comment]
@simonbyrne cool :100:

**Comment by simonbyrne** [comment]
Okay, I've got it working fairly reliably now. See ce059fe0e30b3ee510f6981ca7d60a953f50291d, which is part of #59.

**Comment by simonbyrne** [comment]
Fixed in #59.
